lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp

```diff
@@ -444,6 +444,45 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
                       binder.op, resultType, lhs, rhs);
                   return success();
                 });
+  patterns.onOp(
+      "NegativeLogLikelihoodLoss", 13,
+      [](OpBinder binder, ConversionPatternRewriter &rewriter) {
```

Comment:
> This implements the Onnx.NegativeLogLikelihoodLoss op using the signature provided [here](https://onnx.ai/onnx/operators/onnx__NegativeLogLikelihoodLoss.html) by replacing it with a NLLLossForward op.

This "NegativeLogLikelihoodLoss" op is a "function" in ONNX - do we want this bespoke conversion, or should we expand up in ONNX instead? https://github.com/llvm/torch-mlir/issues/3384 (fyi @andfau-amd)

Sample test case:
* function: https://github.com/nod-ai/SHARK-TestSuite/blob/main/iree_tests/onnx/node/generated/test_nllloss_NC/model.mlir
* expanded: https://github.com/nod-ai/SHARK-TestSuite/blob/main/iree_tests/onnx/node/generated/test_nllloss_NC_expanded/model.mlir

Reply:
It seems this is a "context-dependent" function, which I hadn't yet added support for in my WIP patch for #3384. I added support for it now, and it seems like it produces the right expansion.

I won't comment on whether or not my approach is better. I notice that the result after conversion is different from what this PR would produce; it doesn't use the torch.aten.nll_loss_forward op.

Reply:
"Context-dependent" means the body of the function differs depending on what the call-site looks like. In this case: https://github.com/onnx/onnx/blob/093a8d335a66ea136eb1f16b3a1ce6237ee353ab/onnx/defs/math/defs.cc#L2161-L2271.

Reply:
@andfau-amd @ScottTodd so given that we have a method to expand context-dependent functions in ONNX right now, what should we do about this PR? Are there any other considerations we need to make? I'm not fully aware of the implications of the differing conversions as far as numerics etc.

Reply:
The questions to ask are:
* Do we want a lowering path for this op that is different from the existing upstream expansion?
* Do we want different lowering paths in different contexts, e.g. for different backends (IREE llvm-cpu, IREE spir-v, other torch-mlir users, etc.)

I don't have answers to those questions myself, but I generally lean towards writing less code ourselves and leaning on upstream support whenever possible. Every line of code downstream is another line of code that downstream has to review, test, maintain, build into packages, etc.